browser/static/js/actions.js

```diff
@@ -0,0 +1,179 @@
+class ToggleEdit extends Action {
+
+    do() {
+      edit_mode = !edit_mode;
+    }
+
+    undo() {
+      this.do()
+    }
+
+    redo() {
+      this.do()
+    }
+  }
+
+class ToggleHighlight extends Action {
+
+  do() {
+    current_highlight = !current_highlight;
+  }
+
+  undo() {
+    this.do()
+  }
+
+  redo() {
+    this.do()
+  }
+}
+
+/** Action to change the viewed frame. */
+class ChangeFrame extends Action {
+  constructor(mode, frame) {
+    super();
+    this.mode = mode;
+    this.oldValue = current_frame;
+    // frame mod max_frames
+    this.newValue = ((frame % max_frames) + max_frames) % max_frames;
```

Comment:
`%` is [remainder](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Remainder) in Javascript, not mod and it can be negative, e.g. `-1 % 10 == -1` rather than 9. This expression implements mod in terms of remainder, taken from this [StackOverflow answer](https://stackoverflow.com/questions/4467539/javascript-modulo-gives-a-negative-result-for-negative-numbers).